// powercord/constants.ts
import { join } from "path";

// Replugged
export const WEBSITE = "https://replugged.dev";
export const I18N_WEBSITE = "https://i18n.powercord.dev/projects/powercord";
export const REPO_URL = "replugged-org/replugged";

// Runtime
export const SETTINGS_FOLDER = join(__dirname, "..", "settings");
export const CACHE_FOLDER = join(__dirname, "..", ".cache");
export const LOGS_FOLDER = join(__dirname, "..", ".logs");

// Discord Server
export const DISCORD_INVITE = "B2TcnXV9Rg";
export const GUILD_ID = "1000926524452647132";
export const SpecialChannels = Object.freeze({
  // Support-related
  KNOWN_ISSUES: "1000955953618899084",
  SUPPORT_INSTALLATION: "1000955958442332220",
  SUPPORT_PLUGINS: "1000955957481852968",
  SUPPORT_MISC: "1000955959797088407",

  // Community content
  STORE_PLUGINS: ["1000955968592552047", "755005584322854972"],
  STORE_THEMES: ["1000955969657917551", "755005710323941386"],
  CSS_SNIPPETS: ["1000955971650195588"],
  JS_SNIPPETS: ["1000955970681319494"],
});

// Plugins
export const CORE_PLUGINS = [
  "pc-commands",
  "pc-i18n",
  "pc-moduleManager",
  "pc-notices",
  "pc-settings",
  "pc-updater",
];
