// powercord/constants.js
const { join } = require('path');

module.exports = Object.freeze({
  // Replugged
  WEBSITE: 'https://replugged.dev',
  I18N_WEBSITE: 'https://i18n.replugged.dev/projects/replugged/',
  REPO_URL: 'replugged-org/replugged',

  // Runtime
  SETTINGS_FOLDER: join(__dirname, '..', '..', '..', 'user', 'settings'),
  PLUGINS_FOLDER: join(__dirname, '..', '..', '..', 'user', 'plugins'),
  THEMES_FOLDER: join(__dirname, '..', '..', '..', 'user', 'themes'),
  CACHE_FOLDER: join(__dirname, '..', '..', '..', '.cache'),
  LOGS_FOLDER: join(__dirname, '..', '..', '..', '.logs'),
  CSS_FILE: join(__dirname, '..', '..', '..', 'user', 'themes', 'quickcss.css'),


  // Discord Server
  DISCORD_INVITE: 'B2TcnXV9Rg',
  GUILD_ID: '1000926524452647132',
  SpecialChannels: Object.freeze({
    // Support-related
    KNOWN_ISSUES: '1000955953618899084',
    SUPPORT_INSTALLATION: '1000955958442332220',
    SUPPORT_PLUGINS: '1000955957481852968',
    SUPPORT_MISC: '1000955959797088407',

    // Community content
    STORE_PLUGINS: [ '1000955968592552047', '755005584322854972' ],
    STORE_THEMES: [ '1000955969657917551', '755005710323941386' ],
    CSS_SNIPPETS: [ '1000955971650195588', '755005803303403570' ],
    JS_SNIPPETS: [ '1000955970681319494', '755005784999329883' ]
  })
});
