// powercord/util/index.js
if (!globalThis.structuredClone) {
  /**
   * Add support of structuredClone for NodeJS < 17
   * 
   * Docs: https://developer.mozilla.org/en-US/docs/Web/API/structuredClone
   * 
   * @param {object} obj - object to clone
   * @param {object} _ - (not used)
   * 
   * @return {object} cloned object
   */
  function structuredClone(value, _) {
    return JSON.parse(JSON.stringify(value));
  };
  globalThis.structuredClone = structuredClone;

}

module.exports = {
  sleep: require('./sleep'),
  waitFor: require('./waitFor'),
  wrapInHooks: require('./wrapInHooks'),
  createElement: require('./createElement'),
  getOwnerInstance: require('./getOwnerInstance'),
  getReactInstance: require('./getReactInstance'),
  forceUpdateElement: require('./forceUpdateElement'),
  findInTree: require('./findInTree'),
  findInReactTree: require('./findInReactTree'),
  camelCaseify: require('./camelCaseify'),
  formatTime: require('./formatTime'),
  rmdirRf: require('./rmdirRf'),
  gotoOrJoinServer: require('./gotoOrJoinServer'),
  injectContextMenu: require('./injectContextMenu')
};
