// powercord/entities/Updatable.ts
import Events from 'events';
import { join } from 'path';
import { existsSync } from 'fs';
import { head } from 'powercord/http';

const TIMEOUT = 10e3;

/**
 * @property {String} entityID
 * @property {String} entityPath
 * @property {String} updateIdentifier
 */
class Updatable extends Events {
  entityID: string;
  entityPath: string;
  updateIdentifier: string;
  basePath: string;

  __shortCircuit = false;

  constructor (basePath: string, entityID: string, updateIdentifier: string) {
    super();
    this.basePath = basePath;
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    if (!this.entityID) {
      // It might be pre-defined by plugin manager
      this.entityID = entityID;
    }
    this.entityPath = join(this.basePath, this.entityID);
    if (!updateIdentifier) {
      updateIdentifier = `${this.basePath.split(/[\\/]/).pop()}_${this.entityID}`;
    }
    this.updateIdentifier = updateIdentifier;

    this.__migrateIfNecessary();
  }

  /**
   * @returns {Boolean} Whether this can be updated or not
   */
  isUpdatable (): boolean {
    return existsSync(join(this.basePath, this.entityID, '.git')) && !this.__shortCircuit;
  }

  async _checkForUpdates () {
    try {
      await PowercordNative.exec('git fetch', {
        cwd: this.entityPath,
        timeout: TIMEOUT
      });
      const gitStatus = await PowercordNative.exec('git status -uno', {
        cwd: this.entityPath,
        timeout: TIMEOUT
      }).then(({ stdout }) => stdout!.toString());

      return gitStatus.includes('git pull');
    } catch (e) {
      return false;
    }
  }

  async _getUpdateCommits () {
    const branch = await this.getBranch();
    const commits: {id?: string, author?: string, message?: string}[] = [];

    const gitLog = await PowercordNative.exec(`git log --format="%H -- %an -- %s" ..origin/${branch}`, {
      cwd: this.entityPath,
      timeout: TIMEOUT
    }).then(({ stdout }) => stdout!.toString());

    const lines = gitLog.split('\n');
    lines.pop();
    lines.forEach(line => {
      const data = line.split(' -- ');
      commits.push({
        id: data.shift(),
        author: data.shift(),
        message: data.shift()
      });
    });

    return commits;
  }

  async _update (force = false) {
    try {
      let command = 'git pull --ff-only';
      if (force) {
        const branch = await this.getBranch();
        command = `git reset --hard origin/${branch}`;
      }
      await PowercordNative.exec(command, { cwd: this.entityPath }).then(({ stdout }) => stdout!.toString());
      return true;
    } catch (e) {
      return false;
    }
  }

  /**
   * Fetches the git repository for this entity
   * @returns {Promise<String|null>}
   */
  async getGitRepo (): Promise<string | null> {
    try {
      return await PowercordNative.exec('git remote get-url origin', {
        cwd: this.entityPath,
        timeout: TIMEOUT
      }).then((r) => r.stdout?.toString().match(/github\.com[:/]([\w-_]+\/[\w-_]+)/)?.[1] ?? null);
    } catch (e) {
      console.warn('Failed to fetch git origin url; ignoring.');
      return null;
    }
  }

  /**
   * Fetches the current branch for this entity
   * @returns {Promise<String|null>}
   */
  getBranch (): Promise<string | null> {
    return PowercordNative.exec('git branch', {
      cwd: this.entityPath,
      timeout: TIMEOUT
    }).then(({ stdout }) => stdout?.toString().split('\n').find(l => l.startsWith('*'))?.slice(2).trim() ?? null);
  }

  async __migrateIfNecessary () {
    if (!this.isUpdatable()) {
      return;
    }

    const repo = await this.getGitRepo();
    if (!repo) {
      return;
    }

    const url = `https://github.com/${repo}`;
    const newUrl = await this.__followRedirects(url);
    if (!newUrl) {
      this.__shortCircuit = true;
    } else if (url !== newUrl) {
      console.debug('[Updater] Migrating %s to repository %s', this.entityID, newUrl);
      await PowercordNative.exec(`git remote set-url origin "${newUrl}"`, { cwd: this.entityPath });
    }
  }

  async __followRedirects (url: string) {
    let code = -1;
    do {
      try {
        const res = await head(url);
        code = res.statusCode;
        if (code === 301 || code === 302) {
          if (!res.headers.location) {
            return false;
          }

          url = res.headers.location;
        }
      } catch (e) {
        return false;
      }
    } while (code === 301 || code === 302);
    return url;
  }
}

export default Updatable;
