// powercord/components/settings/index.js
const { getModuleByDisplayName } = require('powercord/webpack');
const AsyncComponent = require('../AsyncComponent');

exports['ButtonItem'] = require('./ButtonItem');
exports['Category'] = require('./Category');
exports['CheckboxInput'] = require('./CheckboxInput');
exports['ColorPickerInput'] = require('./ColorPickerInput');
exports['FormItem'] = require('./FormItem');
exports['KeybindRecorder'] = require('./KeybindRecorder');
exports['RadioGroup'] = require('./RadioGroup');
exports['SelectInput'] = require('./SelectInput');
exports['SliderInput'] = require('./SliderInput');
exports['TextAreaInput'] = require('./TextAreaInput');
exports['TextInput'] = require('./TextInput');

// Add some util components
Object.assign(exports, {
  SwitchItem: AsyncComponent.from(getModuleByDisplayName('SwitchItem'))
});

// Re-export module properties
(async () => {
  const SwitchItem = await getModuleByDisplayName('SwitchItem', true, true);
  [ 'Sizes', 'Themes' ].forEach(prop => exports.SwitchItem[prop] = SwitchItem[prop]);
})();
